src/pushsource/_impl/backend/koji_source.py

```diff
@@ -509,6 +521,94 @@ def _push_items_from_container_build(self, nvr, meta):
 
         return out
 
+    def _push_items_from_vmi_build(self, nvr, meta):
+        LOG.debug("Looking for virtual machine images on %s, %s", nvr, meta)
+
+        if not meta:
+            message = "Virtual machine image build not found in koji: %s" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        # VMI has the {'typeinfo': {'image':{}}} on extra
+        extra = meta.get("extra") or {}
+        typeinfo = extra.get("typeinfo") or {}
+        image = typeinfo.get("image")
+
+        if image is None:
+            message = "Build %s not recognized as a virtual machine image build" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        build_id = meta["id"]
+
+        archives = self._get_archives(build_id)
+
+        # Supported types of VMI archives
+        ami_types = ["raw", "raw-xz"]
+        azure_types = ["vhd", "vhd-compressed"]
+        cloud_types = {}
+
+        for k in ami_types:
+            cloud_types.setdefault(k, AmiPushItem)
+
+        for k in azure_types:
+            cloud_types.setdefault(k, VHDPushItem)
+
+        # Collect all image archives which are not containers (tar)
+        vmi_archives = [
+            elem
+            for elem in archives
+            if elem.get("btype") == "image" and elem.get("type_name") != "tar"
+        ]
+
+        out = []
+
+        # Generate the respective PushItem type for each archive
+        for archive in vmi_archives:
+            path = self._pathinfo.typedir(meta, archive["btype"])
+            item_src = os.path.join(path, archive["filename"])
+
+            extra = archive.get("extra") or {}
+            image = extra.get("image") or {}
+            arch = image.get("arch")
+
+            klass = cloud_types.get(archive["type_name"]) or VMIPushItem
+
+            rel_klass = AmiRelease if klass == AmiPushItem else VMIRelease
+
+            splitted_name = meta["name"].split("-")
+            if len(splitted_name) > 1:
+                product = "-".join(splitted_name[:-1])
+            else:
+                product = splitted_name[0]
+
+            release = rel_klass(
+                arch=arch,
+                date=meta.get("completion_time", "").split(" ")[0],
+                product=product.upper(),
```

Comment:
Where does the requirement to be uppercase come from?

The main reason I ask is that it doesn't seem like this would be koji-specific.  Maybe it would be better for this conversion to happen on the VMIRelease class so that you can be confident it's always uppercase no matter how the object was constructed?

Reply:
Originally I thought we would be using the data from release during some steps for publishing but it seems that this is just used for RHSM which isn't something we will handle in our workflows.

Anyway, for our PoV the product should be like "RHEL" or "RHCOS" so it comes as lowercase from koji (like "rhel-azure") and then got just the first part upper case.

I liked the suggestion to do the uppercase conversion in VMIRelease so I'll adjust it there, thanks for the suggestion

Reply:
Just a note here while I was trying to adjust this:

By using an uppercase converter in the attribute `product` from VMIRelease it breaks the tests `test_source_against_baseline[staged-simple-ami-bc, staged-simple-ami]` with:

```
E       AssertionError: Output differs from baseline:
E       
E       --- 
E       +++ 
E       @@ -29,7 +29,7 @@
E              base_product: null
E              base_version: null
E              date: 2020-05-11
E       -      product: Fake-Product
E       +      product: FAKE-PRODUCT
E              respin: 1
E              type: ga
E              variant: Fake-Variant
```

I even tried to redefine the attribute product just for AmiRelease but that doesn't seem to work. With this I'm letting the uppercase to be in KojiSource for a while until we got a better solution